diff --git a/bt_nv.cpp b/bt_nv.cpp
index 38a5d8c..417807a 100755
--- a/bt_nv.cpp
+++ b/bt_nv.cpp
@@ -88,7 +88,7 @@ typedef struct _nv_items_tag {
 
 nv_items nvItems[3] =
 {
-  {NV_BD_ADDR_I,                 NV_BD_ADDR_SIZE,     NV_WRITE_ONCE_ENABLE},
+  {NV_BD_ADDR_I,                 NV_BD_ADDR_SIZE,     NV_WRITE_ONCE_DISABLE},
   {NV_BT_SOC_REFCLOCK_TYPE_I,    NV_REF_CLOCK_SIZE,   NV_WRITE_ONCE_DISABLE},
   {NV_BT_SOC_CLK_SHARING_TYPE_I, NV_REF_CLOCK_T_SIZE, NV_WRITE_ONCE_DISABLE},
 };
