From 144437e5851fedf32d925e58b1f55a0e8a22cc9f Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Tue, 1 Jul 2014 09:04:20 -0700
Subject: [PATCH] Change persistence path to /mnt/persist

---
 bt_nv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bt_nv.h b/bt_nv.h
index 0046ef2..525c0e0 100755
--- a/bt_nv.h
+++ b/bt_nv.h
@@ -34,7 +34,7 @@ extern "C" {
 #define TRUE 1
 #define FALSE 0
 #define NV_BD_ADDR_SIZE 6
-#define PERSISTENCE_PATH "/persist"
+#define PERSISTENCE_PATH "/mnt/persist"
 #define BT_NV_FILE_NAME ".bt_nv.bin"
 
 /* BTS Master Clock Reference Value */
-- 
1.7.11.4

