From e7491f1388edcf80923acc68ab8eb234bd02d562 Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Thu, 10 Jul 2014 15:04:02 -0700
Subject: [PATCH] btnvtool print mac addr bytes in correct order

---
 btnvtool.cpp | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/btnvtool.cpp b/btnvtool.cpp
index 4bc0d8c..7ae3bac 100644
--- a/btnvtool.cpp
+++ b/btnvtool.cpp
@@ -204,12 +204,12 @@ int main(int argc, char *argv[])
        case 'p':
          bt_nv_cmd(NV_READ_F, NV_BD_ADDR_I, &my_nv_item);
          fprintf (stderr, "--board-address: %x.%x.%x.%x.%x.%x \n",
-                     my_nv_item.bd_addr[0],
-                     my_nv_item.bd_addr[1],
-                     my_nv_item.bd_addr[2],
-                     my_nv_item.bd_addr[3],
+                     my_nv_item.bd_addr[5],
                      my_nv_item.bd_addr[4],
-                     my_nv_item.bd_addr[5]);
+                     my_nv_item.bd_addr[3],
+                     my_nv_item.bd_addr[2],
+                     my_nv_item.bd_addr[1],
+                     my_nv_item.bd_addr[0]);
          bt_nv_cmd(NV_READ_F, NV_BT_SOC_REFCLOCK_TYPE_I, &my_nv_item);
          fprintf (stderr, "--reference-clock: %s MHz \n",
                      my_nv_item.bt_soc_refclock_type?"19.2":"32" );
-- 
1.7.11.4

