From fe1fbaaec3bfca47cf95d9368e27b3eb6ce15515 Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Fri, 11 Jul 2014 08:32:57 -0700
Subject: [PATCH] Print leading 0's when display mac addr

---
 btnvtool.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/btnvtool.cpp b/btnvtool.cpp
index 7ae3bac..fcb9801 100644
--- a/btnvtool.cpp
+++ b/btnvtool.cpp
@@ -203,7 +203,7 @@ int main(int argc, char *argv[])
 
        case 'p':
          bt_nv_cmd(NV_READ_F, NV_BD_ADDR_I, &my_nv_item);
-         fprintf (stderr, "--board-address: %x.%x.%x.%x.%x.%x \n",
+         fprintf (stderr, "--board-address: %02x.%02x.%02x.%02x.%02x.%02x \n",
                      my_nv_item.bd_addr[5],
                      my_nv_item.bd_addr[4],
                      my_nv_item.bd_addr[3],
-- 
1.7.11.4

