From b6fd987d67f1ef50058ed3cf5540a2eb78e6bcd6 Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Wed, 9 Jul 2014 14:01:28 -0700
Subject: [PATCH] Fix include for ptt_socket_app

---
 include/diaglogi.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/diaglogi.h b/include/diaglogi.h
index f7aa428..bc4613e 100755
--- a/include/diaglogi.h
+++ b/include/diaglogi.h
@@ -27,7 +27,7 @@ when       who     what, where, why
 
 
 #include "log_codes.h"
-#include "./../include/log.h"
+#include "log.h"
 // COMMENTED OUT FOR LINUX
 //#include "diagi.h"
 /* -------------------------------------------------------------------------
-- 
1.7.11.4

