diff --git a/mdlog/diag_mdlog.c b/mdlog/diag_mdlog.c
index 34a53d9..6ed25db 100755
--- a/mdlog/diag_mdlog.c
+++ b/mdlog/diag_mdlog.c
@@ -98,7 +98,7 @@ static void usage(char *progname)
 	printf("\n-s, --size:\t maximum file size in MB\n");
 	printf("\n-t, --text:\t print logging to text file\n");
 	printf("\n-x, --terminal:\t print logging to terminal\n");
-	printf("\n-r, --rotate:\t collect log  in rotation at \sdcard\diag_logs\rotate\diag_logs.qmdl \n");	
+	printf("\n-r, --rotate:\t collect log  in rotation at \\sdcard\diag_logs\\rotate\\diag_logs.xxx \n");	
 	printf("\n-k, --kill:\t kill existing instance of diag_mdlog\n");
 	printf("\n-c  --cleanmask:\t Send mask cleanup to modem at exit\n");
 	printf("\n-h, --help:\t usage help\n");
@@ -332,7 +332,7 @@ int main(int argc, char *argv[])
 			(void)strlcat(output_dir, buffer, FILE_NAME_LEN);	
 		}
 		
-		mkdir(output_dir, 0770);		    
+		mkdir(output_dir, 0755);		    
 		if ( errno == 0 || errno == EEXIST)
 		{
 		  DIAG_LOGE("diag_mdlog: logging in directory %s\n", output_dir);
diff --git a/src/diag_lsm.c b/src/diag_lsm.c
index 52255ad..98ded85 100755
--- a/src/diag_lsm.c
+++ b/src/diag_lsm.c
@@ -1329,7 +1329,7 @@ void log_to_device(unsigned char *ptr, int logging_mode, int size)
 			
 				
 						
-			fd_md = open(file_name, O_CREAT | O_RDWR | O_SYNC | O_APPEND, 0644);
+			fd_md = open(file_name, O_CREAT | O_RDWR | O_SYNC | O_APPEND, 0755 );
 			fd_dev = fd_md;
 			if (fd_md < 0) {
 				DIAG_LOGE(" File open error, please check %s ", file_name);
@@ -1366,7 +1366,7 @@ void log_to_device(unsigned char *ptr, int logging_mode, int size)
 						 timestamp_buf, ".txt");
 				}
 					
-				fd_md_txt = open(file_name_txt, O_CREAT | O_RDWR | O_SYNC | O_APPEND, 0644);
+				fd_md_txt = open(file_name_txt, O_CREAT | O_RDWR | O_SYNC | O_APPEND, 0755 );
 				fd_dev_txt = fd_md_txt;
 				if (fd_md_txt < 0) {
 					DIAG_LOGE(" File open error, please check");
