From 4a7cd77f047c999ca4bfd301c9ac87fbc61d45d6 Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Thu, 13 Nov 2014 18:11:42 -0800
Subject: [PATCH 1/1] include stdint.h

---
 gemini/inc/gemini_app_util_mmap.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/gemini/inc/gemini_app_util_mmap.h b/gemini/inc/gemini_app_util_mmap.h
index 794f9ee..daa45e7 100644
--- a/gemini/inc/gemini_app_util_mmap.h
+++ b/gemini/inc/gemini_app_util_mmap.h
@@ -4,6 +4,7 @@
 
 #ifndef __TEST_UTIL_MMAP_H__
 #define __TEST_UTIL_MMAP_H__
+#include <stdint.h>
 #include <linux/msm_ion.h>
 typedef enum {
   GEMINI_PMEM_ADSP,
-- 
1.7.11.4

