diff --git a/server/Android.mk b/server/Android.mk
index 3ea1152..7b35126 100755
--- a/server/Android.mk
+++ b/server/Android.mk
@@ -9,6 +9,32 @@ ifeq ($(FEATURE_FACE_PROC),true)
   LOCAL_MODULE_TAGS := optional eng
   include $(BUILD_MULTI_PREBUILT)
 endif
+
+#************* libcameracsiphyparams starts************#
+LOCAL_PATH:= $(LOCAL_DIR_PATH)
+include $(CLEAR_VARS)
+LOCAL_SRC_FILES:= \
+        hardware/sensor/cust_sens/libcustcamsens_params.c
+
+
+
+LOCAL_SHARED_LIBRARIES:= libcutils libdl
+LOCAL_MODULE:= libcustcamsens_params
+LOCAL_MODULE_TAGS := optional
+include $(BUILD_SHARED_LIBRARY)
+#************* libcameracsiphyparams End************#
+
+#************* exporting cust_sens_params.conf starts************#
+LOCAL_PATH:= $(LOCAL_DIR_PATH)
+include $(CLEAR_VARS)
+LOCAL_MODULE := cust_sens_params.conf
+LOCAL_MODULE_TAGS := optional
+LOCAL_MODULE_CLASS := ETC
+LOCAL_SRC_FILES := hardware/sensor/cust_sens/cust_sens_params.conf
+LOCAL_MODULE_PATH := $(TARGET_OUT_ETC)
+include $(BUILD_PREBUILT)
+#************* exporting cust_sens_params.conf End************#
+
 #************* libmmcamera_target Start ************#
 
 LOCAL_PATH:= $(LOCAL_DIR_PATH)
@@ -103,7 +129,7 @@ endif
 
 # This has to be removed. VERY VERY BAD
 LOCAL_C_INCLUDES+= $(shell find $(LOCAL_PATH) -type d -print )
-LOCAL_SHARED_LIBRARIES:= libC2D2 libcutils libdl
+LOCAL_SHARED_LIBRARIES:= libC2D2 libcutils libdl libcustcamsens_params
 
 LOCAL_COPY_HEADERS_TO := mm-camera
 LOCAL_COPY_HEADERS := ../common/camera_defs_i.h
