diff --git a/server/Makefile.am b/server/Makefile.am
index e54bf48..7082929 100644
--- a/server/Makefile.am
+++ b/server/Makefile.am
@@ -413,25 +413,6 @@ libmmcamera_statsproc31_la_CFLAGS += -I hardware/actuator/0208
 libmmcamera_statsproc31_la_CFLAGS += -I chromatix/$(CHROMATIX_VERSION)
 
 libmmcamera_statsproc31_la_SOURCES := statsproc/stats_proc.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/aec/aec_interface.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/aec/aec.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/aec/aec_util.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/aec/aec_slow_conv.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/aec/aec_snapshot.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/awb/awb_interface.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/awb/awb.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/awb/awb_util.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/awb/awb_self_cal.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/awb/awb_agw.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/asd/asd_interface.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/asd/asd.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_interface.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_util.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_exhstv.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_cont.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_hillclimb.c
-libmmcamera_statsproc31_la_SOURCES += statsproc/af/af_slopepredictive.c
 libmmcamera_statsproc31_la_SOURCES += statsproc/is/dis/dis_interface.c
 libmmcamera_statsproc31_la_SOURCES += statsproc/is/eis/eis_interface.c
 libmmcamera_statsproc31_la_SOURCES += statsproc/is/is_interface.c
@@ -443,6 +424,49 @@ libmmcamera_statsproc31_la_LIBADD = -lmmcamera_image_stab
 libmmcamera_statsproc31_la_LDFLAGS := $(DEBUG_LDFLAGS)
 libmmcamera_statsproc31_la_LDFLAGS += -shared
 
+#libmmcamera_3a_legacy
+lib_LTLIBRARIES += libmmcamera_3a_legacy.la
+
+libmmcamera_3a_legacy_la_CFLAGS := $(DEBUG_CFLAGS)
+libmmcamera_3a_legacy_la_CFLAGS += -DAMSS_VERSION=$(AMSS_VERSION)
+libmmcamera_3a_legacy_la_CFLAGS += -DMSM_CAMERA_GCC
+libmmcamera_3a_legacy_la_CFLAGS += -DLINUX_ENABLED
+libmmcamera_3a_legacy_la_CFLAGS += -D_V4L2_BASED_CAM_
+#remove when snapshot is supported
+#libmmcamera_3a_legacy_la_CFLAGS += -DDISABLE_JPEG_ENCODING
+libmmcamera_3a_legacy_la_CFLAGS += -DUSE_ION
+libmmcamera_3a_legacy_la_CFLAGS += -I ../common
+libmmcamera_3a_legacy_la_CFLAGS += -I statsproc/aec
+libmmcamera_3a_legacy_la_CFLAGS += -I statsproc
+libmmcamera_3a_legacy_la_CFLAGS += -I statsproc/is/dis
+libmmcamera_3a_legacy_la_CFLAGS += -I statsproc/is/eis
+libmmcamera_3a_legacy_la_CFLAGS += -I include
+libmmcamera_3a_legacy_la_CFLAGS += -I hardware/actuator/0208
+libmmcamera_3a_legacy_la_CFLAGS += -I chromatix/$(CHROMATIX_VERSION)
+
+libmmcamera_3a_legacy_la_SOURCES := statsproc/aec/aec_interface.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/aec/aec.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/aec/aec_util.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/aec/aec_slow_conv.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/aec/aec_snapshot.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/awb/awb_interface.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/awb/awb.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/awb/awb_util.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/awb/awb_self_cal.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/awb/awb_agw.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/asd/asd_interface.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/asd/asd.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_interface.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_util.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_exhstv.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_cont.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_hillclimb.c
+libmmcamera_3a_legacy_la_SOURCES += statsproc/af/af_slopepredictive.c
+
+libmmcamera_3a_legacy_la_LDFLAGS := $(DEBUG_LDFLAGS)
+libmmcamera_3a_legacy_la_LDFLAGS += -shared
+
 #libmmcamera_frameproc
 lib_LTLIBRARIES += libmmcamera_frameproc.la
 
