diff --git a/server/hardware/ispif/ispif_interface.c b/server/hardware/ispif/ispif_interface.c
index e523fc7..c127e04 100755
--- a/server/hardware/ispif/ispif_interface.c
+++ b/server/hardware/ispif/ispif_interface.c
@@ -464,14 +464,14 @@ if( !isBayer )
       /*ispif_obj->ispif_ctrl[intftype].ispif_params.csid = sensor_get.data.sensor_csi3d_params[i]->csi_lane_params.csid_core[0];*/
       if (intftype == PIX0) {
 		ispif_obj->ispif_ctrl[intftype].ispif_params.csid = 
-			sensor_get.data.sensor_csi3d_params[1]->csi_lane_params.csid_core[0];
+			sensor_get.data.sensor_csi3d_params[0]->csi_lane_params.csid_core[0];
 	  	/* configure 3d output params */
       	ispif_obj->ispif_ctrl[intftype].output_param.is3D = 1; 
       	ispif_obj->ispif_ctrl[intftype].output_param.x_output_size = 0x668;
 	  }
       else if (intftype == PIX1)
 		ispif_obj->ispif_ctrl[intftype].ispif_params.csid = 
-			sensor_get.data.sensor_csi3d_params[0]->csi_lane_params.csid_core[0];
+			sensor_get.data.sensor_csi3d_params[1]->csi_lane_params.csid_core[0];
 	} else {
       CDBG_ERROR("%s, ispif_params config sensor in non-3d mode",__func__);
       ispif_obj->ispif_ctrl[intftype].ispif_params.csid = sensor_get.data.sensor_csi_params->csi_lane_params.csid_core[0];
