From 0bb9115ce14eb690dec9b370c4832015245383b5 Mon Sep 17 00:00:00 2001
From: sswami <sswami@qti.qualcomm.com>
Date: Mon, 13 Apr 2015 19:14:52 -0700
Subject: [PATCH] fix compilation error with undefined UINT32_MAX

---
 src/adsprpcd.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/adsprpcd.c b/src/adsprpcd.c
index b853c68..9172bf3 100644
--- a/src/adsprpcd.c
+++ b/src/adsprpcd.c
@@ -11,6 +11,8 @@
 #define VERIFY_PRINT_ERROR
 #endif
 
+#define UINT32_MAX 0xFFFFFFFF
+
 #include <stdio.h>
 #include "adsp_default_listener.h"
 #include "verify.h"
-- 
1.7.9.5

