From 0ef8d25faf42a581faf4ae53eafba19b4b03ed7f Mon Sep 17 00:00:00 2001
From: Suraj Swami <sswami@qti.qualcomm.com>
Date: Tue, 19 May 2015 16:38:21 -0700
Subject: [PATCH] use default ADSP_LIBRARY_PATH if envn varaible is not set

---
 inc/apps_std_fopen_env.h |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/inc/apps_std_fopen_env.h b/inc/apps_std_fopen_env.h
index 6f58384..2aeef26 100644
--- a/inc/apps_std_fopen_env.h
+++ b/inc/apps_std_fopen_env.h
@@ -18,9 +18,11 @@ static char* ADSP_LIBRARY_PATH=";/system/lib/rfsa/adsp;/system/vendor/lib/rfsa/a
 #elif (defined __QNX__)
 static char* ADSP_LIBRARY_PATH="/radio/lib/firmware";
 #else
-static char* ADSP_LIBRARY_PATH="";
+static char* ADSP_LIBRARY_PATH="/usr/share/data/adsp";
 #endif
 
+
+
 static __inline int apps_std_fopen_with_env_imp(const char* envvarname, const char* delim, const char* name, const char* mode, apps_std_FILE* psout) {
 #define     EMTPY_STR      ""
 #define     ENV_LEN_GUESS  256
-- 
1.7.9.5

